configure.ac

```diff
@@ -75,8 +74,16 @@ AS_IF([test "x${enable_arch_native}" != "xno"],[
   SQUID_CC_CHECK_ARGUMENT([squid_cv_check_marchnative],[-march=native])
 ])
 
-# might be cross-compiling.
-# NP: BUILDCXXFLAGS defined at the end of configure after CXXFLAGS fully known.
+# If the user did not specify a C++ version.
+user_cxx=`echo "$PRESET_CXXFLAGS" | grep -o -E "(-)std="`
+AS_IF([test "x$user_cxx" = "x"],[
+  # Check for C++17 compiler support
+  # May change CXX.
+  AX_CXX_COMPILE_STDCXX([17],[noext],[mandatory])
+])
+
+# Prerequisite: CXX has been finalized.
+# BUILDCXXFLAGS are defined later, after CXXFLAGS have been finalized.
 AC_ARG_VAR([BUILDCXX],[path to compiler for building compile-time tools. e.g. cf_gen])
```

Comment:
The diff may mislead here because I had to _move_ AX_CXX_COMPILE_STDCXX call from below BUILDCXX-setting code to above BUILDCXX-setting code for cf_gen.cc sources to compile using C++17. This PR does not change BUILDCXX-setting code (which may need more work) and does not change the AS_IF statement guarding AX_CXX_COMPILE_STDCXX. The PR description documents this bug-fixing move.

Reply:
I think this came up downstream at https://bugs.gentoo.org/921688 where the right flags weren't used to build `cf_gen.cc`. We aren't shipping GCC 9 anymore but the user had it installed still.

Reply:
> I think this came up downstream at https://bugs.gentoo.org/921688 where the right flags weren't used to build `cf_gen.cc`.

```
checking whether x86_64-pc-linux-gnu-g++ supports C++17 features with -std=c++17... yes
checking for x86_64-pc-linux-gnu-g++ -std=c++17 ...
...
BUILDCXXFLAGS=-march=core2 -O2
...
libtool: compile:  x86_64-pc-linux-gnu-g++ -std=c++17 ... -march=core2 -O2 -c assert.cc ...
...
x86_64-pc-linux-gnu-g++ -march=core2 -O2 -o cf_gen ./cf_gen.cc ...
```

Yes, the above output looks like a sign of Squid bugs to me.

I hope this cf_gen building bug predates this PR and that this PR changes make fixing that bug a bit easier. The old comment you are responding to here claims that "this PR does not change BUILDCXX-setting code (which may need more work)". The corresponding commit 09835fe message has more on that subject.

It is possible that this bug has been fixed already. I see the expected cf_gen build command in some random build log of v6.6 code:

```
g++ -std=c++17 ... -o cf_gen ./cf_gen.cc
```

However, it is possible that the bug just does not manifest itself in that specific build environment (Ubuntu 22.04).

At any rate, if you can, please file a bug report with Squid [Bugzilla](https://bugs.squid-cache.org/). This old merged PR is not the best place to discuss these issues.